### PR TITLE
✨ new feat: selector middleware

### PR DESCRIPTION
**Pull request by aimuz**
It allows to set check rules to whitelist or blacklist middleware such as Auth
Requests that match the rules will go to the next level of middleware

### CODE REVIEW

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]
This looks cool, but I think it's missing some things:

1. Where are the tests? We should include some tests of this new functionality.
2. Could you include an `selector_example_test.go` file with an example of how to use this? It will automatically be shown as part of the documentation. See https://go.dev/blog/examples.

**Comment by aimuz** [comment]
Yes, that's how it should be. I'll take care of it.

File: interceptors/selector/doc.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/*
+Package selector
+
+`selector` a generic server-side selector middleware for gRPC.
+
+# Server Side Selector Middleware
+It allows to set check rules to whitelist or blacklist middleware such as Auth
```
Comment (johanbrandhorst) [inline interceptors/selector/doc.go:7]: Thanks for the updates, do you mind changing these phrases to `allowlist` and `blocklist` consistently through this PR?
```suggestion
It allows to set check rules to allowlist or blocklist middleware such as Auth
```

File: interceptors/selector/selector.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+type MatchFunc func(ctx context.Context, fullMethod string) bool
+
+type options struct {
+	match MatchFunc
+}
+
+type Option func(o *options)
+
+// WithMatch ...
```
Comment (johanbrandhorst) [inline interceptors/selector/selector.go:17]: I think this comment needs improving
```suggestion
// WithMatch configures what condition should trigger the selector to run the interceptor.
// If used multiple times, the last one specified will be used.
```

File: interceptors/selector/selector.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+func evaluateServerOpt(opts []Option) *options {
+	optCopy := &options{}
+	optCopy.match = AlwaysMatch
+	for _, o := range opts {
+		o(optCopy)
+	}
+	return optCopy
+}
+
+// AlwaysMatch is the default implementation, always selected
```
Comment (johanbrandhorst) [inline interceptors/selector/selector.go:33]:
```suggestion
// AlwaysMatch can be used to always run the interceptor. This is the default behavior.
```

File: interceptors/selector/selector.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+func AlwaysMatch(_ context.Context, _ string) bool {
+	return true
+}
+
+// UnaryServerInterceptor returns a new unary server interceptors that performs per-request selector.
```
Comment (johanbrandhorst) [inline interceptors/selector/selector.go:38]:
```suggestion
// UnaryServerInterceptor returns a new unary server interceptor that will decide whether to call the interceptor on the behavior of the MatchFunc.
```

File: interceptors/selector/selector.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+func UnaryServerInterceptor(interceptors grpc.UnaryServerInterceptor, opts ...Option) grpc.UnaryServerInterceptor {
```
Comment (johanbrandhorst) [inline interceptors/selector/selector.go:39]: I'm leaning towards removing `MatchFunc` as an option and making it a required parameter. This solves two problems:

1. If you accidentally use the selector without a match func, you gain no utility from it.
2. If you accidentally specify two match func options, you might be confused about which to use.

That simplifies things generally, no need for the `AlwaysMatch` function, etc.

File: interceptors/selector/doc.go [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+Requests that match the rules will go to the next level of middleware
```
Comment (johanbrandhorst) [inline interceptors/selector/doc.go:8]:
```suggestion
interceptors to toggle behavior on or off based on the request path.
```

File: interceptors/selector/selector_example_test.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+	}
+
+	ctx = logging.InjectFields(ctx, logging.Fields{"auth.sub", userClaimFromToken(tokenInfo)})
+
+	// WARNING: In production define your own type to avoid context collisions.
+	return context.WithValue(ctx, tokenInfoKey, tokenInfo), nil
+}
+
+func loginSkip(ctx context.Context, fullMethod string) bool {
+	return fullMethod != "login"
```
Comment (johanbrandhorst) [inline interceptors/selector/selector_example_test.go:67]: Lets use a realistic looking path, like you are doing in the tests
```suggestion
	return fullMethod != "/auth.v1.AuthService/Login"
```

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]
Thanks for the tests, just a few suggestions

File: interceptors/selector/selector.go (hunk 5 above)
Comment (aimuz) [reply interceptors/selector/selector.go]: Yes, I readjusted it, can you help me in seeing what is wrong?

File: interceptors/selector/selector.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+	}
+}
+
+// Block the interceptor will not run after the method matches
+func Block(methods []string) MatchFunc {
+	allow := Allow(methods)
+	return func(ctx context.Context, fullMethod string) bool {
+		return !allow(ctx, fullMethod)
+	}
+}
```
Comment (johanbrandhorst) [inline interceptors/selector/selector.go:29]: I don't think we need these, to be honest. Users can easily implement these themselves.

File: interceptors/selector/selector_test.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+import (
+	"context"
+	"github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/auth"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+	"google.golang.org/grpc"
+	"testing"
+)
+
+var whiteList = []string{"/auth.v1beta1.AuthService/Login"}
```
Comment (johanbrandhorst) [inline interceptors/selector/selector_test.go:12]:
```suggestion
var allowList = []string{"/auth.v1beta1.AuthService/Login"}
```

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]

File: interceptors/selector/selector_test.go (hunk 9 above)
Comment (aimuz) [reply interceptors/selector/selector_test.go]: I have adjusted

File: testing/testpb/test.pb.go [hunk 10]
```diff
@@ -1,18 +1,16 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
-// 	protoc-gen-go v1.25.0
+// 	protoc-gen-go v1.28.0
```
Comment (johanbrandhorst) [inline testing/testpb/test.pb.go:3]: Can you revert these changes, I don't think these should need to be changed?

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]

**Review by johanbrandhorst — APPROVED** [review]

**Comment by johanbrandhorst** [comment]
Looks like there is a formatting issue, can you run `make lint`?

**Comment by aimuz** [comment]
done
